include/revng/ADT/UpcastablePointer.h

```diff
@@ -179,6 +179,11 @@ class UpcastablePointer {
   }
 
 public:
+  UpcastablePointer &operator=(std::nullptr_t) noexcept {
+    Pointer.reset(nullptr);
+    return *this;
+  }
```

Comment:
I understand, but, despite the name, the goal of `UpcastablePointer` kinda is to hide the fact that there's a pointer.
It'd be nice if we didn't offer any additional API w.r.t. `std::variant`.
But maybe this is not such an important goal.

Reply:
I did try to suggest something like `UpcastableReference` when I just started working on this PR. But since we went with the pointer representing nullability of some fields, something like this is necessary. If you want, I can replace usage with `reset`, since there's only a handful of places where this is used, but I feel like this makes it easier to read.